server/pages/index.js

```diff
@@ -390,14 +401,32 @@ function setUpLoggedInRoute( req, res, next ) {
 					}
 
 					console.error( 'API Error: ' + errorMessage );
-
-					next( error );
 				}
 
-				return;
+				return error;
 			} );
+
+		Promise.all( [ langPromise, userPromise ] ).then( results => {
+			const rejectedWithErrors = results.filter( r => null != r && r.error );
```

Comment:
@ockham I especially need your help here. This is just a demonstration of my rough idea and I'm not even sure if it worked or not.

When there is only `user-bootstrap`, it is `next( error )` if something goes wrong. What would we do with the `error` passed in and where is the code handling it? Now that there are multiple async computations that could go wrong, how do you envision us handling them?

Reply:
Good questions! 

> What would we do with the `error` passed in and where is the code handling it?

Passing an object to `next` will cause our server-side router -- Express -- to fall through to an error handling middleware, which is defined by taking 4 arguments `( err, req, res, next )` rather than the usual 3 `( req, res, next )`. In our case, that middleware is defined here:

https://github.com/Automattic/wp-calypso/blob/fcaba1e63aa6cb802cc1842e97fdde1ac4f50e56/server/pages/index.js#L498-L513

Reply:
> This is just a demonstration of my rough idea and I'm not even sure if it worked or not.
> [...]
> Now that there are multiple async computations that could go wrong, how do you envision us handling them?

I haven't honestly used `Promise.all( [ langPromise, userPromise ] )` myself :sweat_smile: 

Looking at the [MDN docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all), it seems to me that we'd only enter the `.then()` clause if all promises in the list are fulfilled -- which would mean that `rejectedWithErrors` would always be empty.

If I'm reading those docs correctly, we need a `catch()` clause instead:

> It rejects with the reason of the first promise that rejects.

Reply:
I think we would call a normal `next()` in the `.then()` clause, and a `next( err )` in the `.catch()` clause.

Reply:
To think about it more, I think you are right that I have gone too far. I've simplified it in af8d712 so all the Promise objects simply `throw` the errors and let the `Promise.all()` catch function do the rest.